src/sqlfluff/rules/L003.py

```diff
@@ -416,358 +475,204 @@ def _process_current_line(
-                    template_block_level += 1
+        if this_line.templated_line_type == "end":
+            return self._handle_template_blocks(
+                this_line=this_line,
+                trigger_segment=trigger_segment,
+                previous_lines=previous_lines,
+                memory=memory,
+            )
+        # Assuming it's not a hanger, let's compare it to the other previous
+        # lines. We do it in reverse so that closer lines are more relevant.
 
-                if template_block_level != 0:
+        def _find_previous_line(
+            this_line: _LineSummary,
+            previous_lines: List[_LineSummary],
+            ignoreable_lines: Set[int],
+        ) -> Optional[_LineSummary]:
+            for prev_line in previous_lines:
+                should_ignore = prev_line.line_no in ignoreable_lines
+                if should_ignore or prev_line.is_empty_line:
                     continue
 
-                # Found prior template block line with the same indent balance.
+                # Work out the difference in indent
+                indent_diff = (
+                    this_line.anchor_indent_balance - prev_line.anchor_indent_balance
+                )
+                # If we're comparing to a previous, more deeply indented line,
+                # then skip and keep looking.
+                if indent_diff < 0:
+                    continue
+                return prev_line
+            return None
 
-                # Is this a problem line?
-                if k in memory["problem_lines"] + memory["hanging_lines"]:
-                    # Skip it if it is
-                    return LintResult(memory=memory)
+        prev_line = _find_previous_line(
+            this_line,
+            previous_lines,
+            memory.ignorable_lines,
+        )
 
-                self.logger.debug("    [template block end] Comparing to #%s", k)
-                if this_line["indent_size"] == res[k]["indent_size"]:
-                    # All good.
-                    return LintResult(memory=memory)
+        if not prev_line:
+            return LintResult(memory=memory)
 
+        prev_line_no = prev_line.line_no
+        indent_diff = this_line.anchor_indent_balance - prev_line.anchor_indent_balance
+        # Is the indent balance the same?
+        if indent_diff == 0:
+            self.logger.debug(
+                "    [same indent balance] Comparing to #%s", prev_line_no
+            )
+            if this_line.indent_size != prev_line.indent_size:
                 # Indents don't match even though balance is the same...
-                memory["problem_lines"].append(this_line_no)
-
-                # The previous indent.
-                desired_indent = "".join(elem.raw for elem in res[k]["indent_buffer"])
-
-                # Make fixes
+                memory.problem_lines.add(this_line_no)
+                desired_indent = "".join(elem.raw for elem in prev_line.indent_buffer)
```

Comment:
This line is very strange - Tests refused to pass when I try to generate it from the various numbers we have (as has been done with other such examples).
Seems that this would respect a mix of tabs and spaces (which I believe is the test that fails) - Why would that be?

Reply:
Can you paste the "not working" code here? Not sure I understand.

Reply:
Sorry I actually figure this out.
```sql
SELECT
\t\s\s\s\s col_1,
col_2 --> [corrects to] ---> \t\s\s\s\s col_2,
```
Reusing the mixed spaces from above was hard coded into one of the tests. As Ive mentioned elsewhere I think this isnt a good idea (adding complexity and compute, for negative benefit)

The test was only flagging / forcing this behaviour on one level of misindent, which lead to the surprising behaviour of line 561 being the only one that couldn't be changed.

New algo always puts fresh spaces or tabs depending on the user preference.
1. Avoids expensive call to .raw
2. removes many branches from the code's algo